### PR TITLE
Mccalluc/notebook anndata

### LINKED ISSUE #2482: expand multi-dataset notebook
We also had discussed a scenario where we would check for a particular assay type at the top of the notebook, e.g., to make a notebook that runs a particular analysis.

Example:
1. from the list of input datasets select the ones that are of type scRNA-seq (salmon) or snRNA-seq (salmon) or CODEX (SPRM), spit out some stats about number of datasets in the input list that were ignored and number that can be used for analysis
1. the aforementioned datasets should all contain information about cells and transcript/antibody quantifications as an anndata Zarr store. Add a block that iterates over all datasets in the filtered list, opens the anndata (it might be in a slightly different place for each assay type, at least for CODEX) and parses out the number of cells for the dataset. 
1. Plot a bar chart with the number of cells for each dataset and maybe group by assay type and tissue (which you could get from your metadata table)

I think this would be a great starting point for anyone to start doing cross-dataset analysis.

_Originally posted by @ngehlenborg in https://github.com/hubmapconsortium/portal-ui/issues/2479#issuecomment-1058516022_

### PR DESCRIPTION
**Pull request by mccalluc**
Filing as draft. Towards: fix #2482.

Right now the last cell in the notebook returns something like:
```
{'59249f23ecdcc975c90de9a1956a5285': [
    'https://assets.hubmapconsortium.org/59249f23ecdcc975c90de9a1956a5285/hubmap_ui/anndata-zarr/scvelo_annotated.zarr',
    'https://assets.hubmapconsortium.org/59249f23ecdcc975c90de9a1956a5285/hubmap_ui/anndata-zarr/secondary_analysis.zarr']}
```

- Right now I'm just looking though the list of files to find `.zarr`... but maybe it should depend on the visualization hints for the assays? Or just hardcode the list of assays we are looking for?
- Are there zarr stores we use that are not AnnData? Can they be distinguished on the basis of filenames?
- Once I have a URL like this, what do I do this this? Can [anndata.read_zarr()](https://anndata.readthedocs.io/en/stable/anndata.read_zarr.html) take a URL?

### CODE REVIEW

**Comment by ilan-gold** [comment]
@mccalluc 

```python
>>> import anndata
>>> anndata.read_zarr('https://assets.hubmapconsortium.org/7cc96210e80d5736ca38711503e4ff06/hubmap_ui/anndata-zarr/secondary_analysis.zarr')
AnnData object with n_obs × n_vars = 0 × 0
```
I will ask though about this.

File: context/app/routes_notebooks.py [hunk 1]
```diff
@@ -50,12 +52,35 @@ def notebook(entity_type):
...
+    # Only for datasets with anndata:
+    uuids_to_zarr_files = _get_uuids_to_zarr_files(uuids)
+    zarr_files = set().union(*uuids_to_zarr_files.values())
+    if zarr_files:
+        cells += get_anndata_cells(uuids_to_zarr_files)
+
     return _nb_response(entity_type, cells)
+
+
+def _get_uuids_to_zarr_files(uuids):
```
Comment (ilan-gold) [inline context/app/routes_notebooks.py:77]: doesn't this actually return the full oath to the zarr files? so in this sense, we are not returning uuids but paths?  maybe i'm misreading the function name or something

**Review by ilan-gold — COMMENTED** [review]
ok yes let's wait to hear back from the anndata folks but looks good.

**Comment by ilan-gold** [comment]
@mccalluc can we try 

```python

>>> import anndata
>>> import zarr
>>> anndata.read_zarr(zarr.open('https://assets.hubmapconsortium.org/7cc96210e80d5736ca38711503e4ff06/hubmap_ui/anndata-zarr/secondary_analysis.zarr'))

**Comment by ilan-gold** [comment]
Evidently this also doesn't work....

**Comment by mccalluc** [comment]
Suggestion from Mark:
[remote_anndata_zarr.ipynb.txt](https://github.com/hubmapconsortium/portal-ui/files/8227201/remote_anndata_zarr.ipynb.txt)

**Comment by mccalluc** [comment]
@keller-mark : Thanks for your notebook. Trying to evaluate it, I get an error:

<details><summary>AttributeError: 'FSMap' object has no attribute 'getitems'</summary>

```
adata = load_remote_anndata_zarr(zarr_url, ["leiden"], [], ["X_umap"], [])
```
```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-4-6ef1da3f00d6> in <module>
----> 1 adata = load_remote_anndata_zarr(zarr_url, ["leiden"], [], ["X_umap"], [])

<ipython-input-2-5f5f4c3d126a> in load_remote_anndata_zarr(zarr_url, obs_cols, var_cols, obsm_keys, varm_keys)
      6     var_index = zarr.open(zarr_url + "/var/_index")
      7 
----> 8     obs_df = pd.DataFrame(columns=obs_cols, index=obs_index)
      9     var_df = pd.DataFrame(columns=var_cols, index=var_index)
     10     for col in obs_cols:

/opt/anaconda3/lib/python3.7/site-packages/pandas/core/frame.py in __init__(self, data, index, columns, dtype, copy)
    527 
    528         elif isinstance(data, dict):
--> 529             mgr = init_dict(data, index, columns, dtype=dtype)
    530         elif isinstance(data, ma.MaskedArray):
    531             import numpy.ma.mrecords as mrecords

/opt/anaconda3/lib/python3.7/site-packages/pandas/core/internals/construction.py in init_dict(data, index, columns, dtype)
    258             index = extract_index(arrays[~missing])
    259         else:
--> 260             index = ensure_index(index)
    261 
    262         # no obvious "empty" int column

/opt/anaconda3/lib/python3.7/site-packages/pandas/core/indexes/base.py in ensure_index(index_like, copy)
   5883         # attribute "name"  [union-attr]
   5884         name = index_like.name  # type: ignore[union-attr, attr-defined]
-> 5885         return Index(index_like, name=name, copy=copy)
   5886 
   5887     if is_iterator(index_like):

/opt/anaconda3/lib/python3.7/site-packages/pandas/core/indexes/base.py in __new__(cls, data, dtype, copy, name, tupleize_cols, **kwargs)
    372             raise cls._scalar_data_error(data)
    373         elif hasattr(data, "__array__"):
--> 374             return Index(np.asarray(data), dtype=dtype, copy=copy, name=name, **kwargs)
    375         else:
    376             if tupleize_cols and is_list_like(data):

/opt/anaconda3/lib/python3.7/site-packages/numpy/core/_asarray.py in asarray(a, dtype, order)
     83 
     84     """
---> 85     return array(a, dtype, copy=False, order=order)
     86 
     87 

/opt/anaconda3/lib/python3.7/site-packages/zarr/core.py in __array__(self, *args)
    456 
    457     def __array__(self, *args):
--> 458         a = self[...]
    459         if args:
    460             a = a.astype(args[0])

/opt/anaconda3/lib/python3.7/site-packages/zarr/core.py in __getitem__(self, selection)
    660 
    661         fields, selection = pop_fields(selection)
--> 662         return self.get_basic_selection(selection, fields=fields)
    663 
    664     def get_basic_selection(self, selection=Ellipsis, out=None, fields=None):

/opt/anaconda3/lib/python3.7/site-packages/zarr/core.py in get_basic_selection(self, selection, out, fields)
    786         else:
    787             return self._get_basic_selection_nd(selection=selection, out=out,
--> 788                                                 fields=fields)
    789 
    790     def _get_basic_selection_zd(self, selection, out=None, fields=None):

/opt/anaconda3/lib/python3.7/site-packages/zarr/core.py in _get_basic_selection_nd(self, selection, out, fields)
    828         indexer = BasicIndexer(selection, self)
    829 
--> 830         return self._get_selection(indexer=indexer, out=out, fields=fields)
    831 
    832     def get_orthogonal_selection(self, selection, out=None, fields=None):

/opt/anaconda3/lib/python3.7/site-packages/zarr/core.py in _get_selection(self, indexer, out, fields)
   1124             lchunk_coords, lchunk_selection, lout_selection = zip(*indexer)
   1125             self._chunk_getitems(lchunk_coords, lchunk_selection, out, lout_selection,
-> 1126                                  drop_axes=indexer.drop_axes, fields=fields)
   1127 
   1128         if out.shape:

/opt/anaconda3/lib/python3.7/site-packages/zarr/core.py in _chunk_getitems(self, lchunk_coords, lchunk_selection, out, lout_selection, drop_axes, fields)
   1834         else:
   1835             partial_read_decode = False
-> 1836             cdatas = self.chunk_store.getitems(ckeys, on_error="omit")
   1837         for ckey, chunk_select, out_select in zip(ckeys, lchunk_selection, lout_selection):
   1838             if ckey in cdatas:

/opt/anaconda3/lib/python3.7/site-packages/zarr/storage.py in getitems(self, keys, **kwargs)
   1083     def getitems(self, keys, **kwargs):
   1084         keys_transformed = [self._normalize_key(key) for key in keys]
-> 1085         results = self.map.getitems(keys_transformed, on_error="omit")
   1086         # The function calling this method may not recognize the transformed keys
   1087         # So we send the values returned by self.map.getitems back into the original key space.

AttributeError: 'FSMap' object has no attribute 'getitems'
```

</details>

Strange that it worked for you... is there some difference in library versions?

<details><summary>pip freeze</summary>

```
aiofiles==0.7.0
alabaster==0.7.12
anaconda-client==1.7.2
anaconda-navigator==1.9.7
anaconda-project==0.8.3
aniso8601==8.0.0
anndata==0.6.22.post1
apipkg==1.5
appdirs==1.4.3
appnope==0.1.0
appscript==1.0.1
asciitree==0.3.3
asn1crypto==1.0.1
astroid==2.3.1
astropy==3.2.2
atomicwrites==1.3.0
attrs==19.2.0
autopep8==1.5.4
Babel==2.7.0
backcall==0.1.0
backports.functools-lru-cache==1.5
backports.os==0.1.1
backports.shutil-get-terminal-size==1.0.0
backports.tempfile==1.0
backports.weakref==1.0.post1
beautifulsoup4==4.8.0
bitarray==1.0.1
bkcharts==0.2
black==22.1.0
bleach==3.1.0
bokeh==1.3.4
boto==2.49.0
boto3==1.14.0
Bottleneck==1.2.1
cached-property==1.5.1
cachetools==5.0.0
cellxgene==0.13.0
certifi==2019.9.11
cffi==1.12.3
chardet==3.0.4
click==8.0.4
click-default-group==1.2.2
cloudpickle==1.2.2
clyent==1.2.2
colorama==0.4.4
conda==4.7.12
conda-build==3.18.9
conda-package-handling==1.6.0
conda-verify==3.4.2
contextlib2==0.6.0
cryptography==2.7
cycler==0.10.0
Cython==0.29.13
cytoolz==0.10.0
dask==2.5.2
datapackage==1.14.1
decorator==4.4.0
defusedxml==0.6.0
distributed==2.5.2
docutils==0.15.2
entrypoints==0.3
et-xmlfile==1.0.1
execnet==1.8.0
fastcache==1.1.0
fasteners==0.16
filelock==3.0.12
flake8==3.8.2
Flask-Caching==1.8.0
Flask-Compress==1.4.0
Flask-Cors==3.0.8
Flask-RESTful==0.3.7
flatbuffers==1.11
fsspec==0.5.2
future==0.17.1
generate-tiff-offsets==0.1.7
gevent==1.4.0
glob2==0.7
globus-sdk==2.0.1
gmpy2==2.0.8
goodtables==2.4.16
greenlet==0.4.15
h11==0.12.0
h2==4.0.0
h5py==2.9.0
HeapDict==1.0.1
hpack==4.0.0
html5lib==1.0.1
hubmap-api-py-client==0.0.9
hubmap-cell-id-gen-py==0.1.0
Hypercorn==0.11.2
hyperframe==6.0.1
idna==2.8
ijson==3.0.4
imageio==2.6.0
imagesize==1.1.0
importlib-metadata==0.23
iniconfig==1.1.1
ipykernel==5.1.2
ipython==7.8.0
ipython-genutils==0.2.0
ipywidgets==7.6.5
isodate==0.6.0
isort==4.3.21
itsdangerous==1.1.0
jdcal==1.4.1
jedi==0.15.1
Jinja2==2.10.3
jmespath==0.9.4
joblib==0.13.2
js-regex==1.0.1
json5==0.8.5
jsonlines==1.2.0
jsonpointer==2.0
jsonref==0.2
jsonschema==3.2.0
jupyter==1.0.0
jupyter-client==5.3.3
jupyter-console==6.0.0
jupyter-core==4.5.0
jupyterlab==1.1.4
jupyterlab-server==1.0.6
jupyterlab-widgets==1.0.2
keyring==18.0.0
kiwisolver==1.1.0
lazy-object-proxy==1.4.2
libarchive-c==2.8
lief==0.9.0
linear-tsv==1.1.0
llvmlite==0.29.0
locket==0.2.0
lxml==4.4.1
Markdown==3.2.2
MarkupSafe==1.1.1
matplotlib==3.1.1
mccabe==0.6.1
mistune==0.8.4
mkl-fft==1.0.14
mkl-random==1.1.0
mkl-service==2.3.0
mock==3.0.5
more-itertools==7.2.0
mpmath==1.1.0
msgpack==0.6.1
multipledispatch==0.6.0
mypy-extensions==0.4.3
natsort==7.0.0
navigator-updater==0.2.1
nbconvert==5.6.0
nbformat==4.4.0
negspy==0.2.24
neo4j==4.4.1
networkx==2.3
nltk==3.4.5
nose==1.3.7
notebook==6.0.1
numba==0.45.1
numcodecs==0.7.3
numexpr==2.7.0
numpy==1.17.2
numpydoc==0.9.1
olefile==0.46
openpyxl==3.0.0
packaging==19.2
pandas==1.2.4
pandocfilters==1.4.2
parso==0.5.1
partd==1.0.0
path.py==12.0.1
pathlib2==2.3.5
pathspec==0.9.0
patsy==0.5.1
pep8==1.7.1
pexpect==4.7.0
pickleshare==0.7.5
Pillow==6.2.0
pkginfo==1.5.0.1
platformdirs==2.5.1
pluggy==0.13.0
ply==3.11
portal-visualization==0.0.1
priority==1.3.0
prometheus-client==0.7.1
prompt-toolkit==2.0.10
property==2.2
prov==2.0.0
psutil==5.6.3
ptyprocess==0.6.0
py==1.10.0
pyasn1==0.4.8
pycodestyle==2.6.0
pycosat==0.6.3
pycparser==2.19
pycrypto==2.6.1
pycurl==7.43.0.3
pyflakes==2.2.0
Pygments==2.8.1
PyJWT==1.7.1
pylint==2.4.2
pyodbc==4.0.27
pyOpenSSL==19.0.0
pyparsing==2.4.2
pyrsistent==0.15.4
PySocks==1.7.1
pytest==6.2.2
pytest-arraydiff==0.3
pytest-astropy==0.5.0
pytest-doctestplus==0.4.0
pytest-forked==1.3.0
pytest-mock==1.11.1
pytest-openfiles==0.4.0
pytest-remotedata==0.3.2
pytest-xdist==2.2.1
python-dateutil==2.8.0
pytz==2021.3
PyWavelets==1.0.3
PyYAML==5.1.2
pyzmq==18.1.0
QtAwesome==0.6.0
qtconsole==4.5.5
QtPy==1.9.0
rdflib==6.1.1
readme-renderer==29.0
requests==2.25.1
requests-toolbelt==0.9.1
rfc3986==1.4.0
rope==0.14.0
rsa==3.4.2
ruamel-yaml==0.15.46
s3transfer==0.3.2
scikit-image==0.15.0
scikit-learn==0.21.3
scipy==1.3.1
seaborn==0.9.0
Send2Trash==1.5.0
simpleeval==0.9.10
simplegeneric==0.8.1
singledispatch==3.4.0.3
six==1.12.0
snowballstemmer==2.0.0
sortedcollections==1.1.2
sortedcontainers==2.1.0
soupsieve==1.9.3
Sphinx==2.2.0
sphinxcontrib-applehelp==1.0.1
sphinxcontrib-devhelp==1.0.1
sphinxcontrib-htmlhelp==1.0.2
sphinxcontrib-jsmath==1.0.1
sphinxcontrib-qthelp==1.0.2
sphinxcontrib-serializinghtml==1.1.3
sphinxcontrib-websupport==1.1.2
spyder==3.3.6
spyder-kernels==0.5.2
SQLAlchemy==1.3.9
starlette==0.14.0
statistics==1.0.3.5
statsmodels==0.10.1
sympy==1.4
tables==3.5.1
tableschema==1.19.2
tabulator==1.52.0
tblib==1.4.0
terminado==0.8.2
testpath==0.4.2
tifffile==2020.10.1
toml==0.10.2
tomli==2.0.1
toolz==0.10.0
tornado==6.0.3
tqdm==4.36.1
traitlets==4.3.3
twine==3.2.0
typed-ast==1.5.2
typing-extensions==3.10.0.0
ujson==4.0.2
unicodecsv==0.14.1
urllib3==1.26.5
virtualenv==20.0.1
vitessce==1.0.5
wcwidth==0.1.7
webencodings==0.5.1
Werkzeug==0.16.0
widgetsnbextension==3.5.1
wrapt==1.11.2
wsproto==1.0.0
wurlitzer==1.0.3
xlrd==1.2.0
XlsxWriter==1.2.1
xlwings==0.15.10
xlwt==1.3.0
yattag==1.12.2
zarr==2.8.3
zict==1.0.0
zipp==0.6.0
```

</details>

**Comment by mccalluc** [comment]
Replaced by #2515

**Pull request closed without merging**